plutarch-orphanage/src/PlutusLedgerApi/V2/Orphans.hs

```diff
@@ -0,0 +1,1271 @@
+      . coarbitrary routs
+      . coarbitrary outs
+      . coarbitrary fee
+      . coarbitrary mint
+      . coarbitrary dcert
+      . coarbitrary wdrl
+      . coarbitrary validRange
+      . coarbitrary sigs
+      . coarbitrary reds
+      . coarbitrary dats
+      . coarbitrary tid
+
+-- | @since 1.0.0
+instance Function PLA.TxInfo where
+  {-# INLINEABLE function #-}
+  function = functionMap into outOf
+    where
+      -- We have to nest tuples as Function doesn't have instances for anything
+      -- bigger than a 6-tuple.
+      into ::
+        PLA.TxInfo ->
+        ( [PLA.TxInInfo]
+        , [PLA.TxInInfo]
+        , [PLA.TxOut]
+        , PLA.Value
+        , PLA.Value
+        , [PLA.DCert]
+        , ( AssocMap.Map PLA.StakingCredential Integer
+          , PLA.POSIXTimeRange
+          , [PLA.PubKeyHash]
+          , AssocMap.Map PLA.ScriptPurpose PLA.Redeemer
+          , AssocMap.Map PLA.DatumHash PLA.Datum
+          , PLA.TxId
+          )
+        )
+      into (PLA.TxInfo ins routs outs fee mint dcert wdrl validRange sigs reds dats tid) =
+        (ins, routs, outs, fee, mint, dcert, (wdrl, validRange, sigs, reds, dats, tid))
+      outOf ::
+        ( [PLA.TxInInfo]
+        , [PLA.TxInInfo]
+        , [PLA.TxOut]
+        , PLA.Value
+        , PLA.Value
+        , [PLA.DCert]
+        , ( AssocMap.Map PLA.StakingCredential Integer
+          , PLA.POSIXTimeRange
+          , [PLA.PubKeyHash]
+          , AssocMap.Map PLA.ScriptPurpose PLA.Redeemer
+          , AssocMap.Map PLA.DatumHash PLA.Datum
+          , PLA.TxId
+          )
+        ) ->
+        PLA.TxInfo
+      outOf (ins, routs, outs, fee, mint, dcert, (wdrl, validRange, sigs, reds, dats, tid)) =
+        PLA.TxInfo ins routs outs fee mint dcert wdrl validRange sigs reds dats tid
+
+-- | @since 1.0.0
+instance Arbitrary PLA.ScriptContext where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = PLA.ScriptContext <$> arbitrary <*> arbitrary
+  {-# INLINEABLE shrink #-}
+  shrink (PLA.ScriptContext txi purpose) = PLA.ScriptContext <$> shrink txi <*> shrink purpose
+
+-- | @since 1.0.0
+instance CoArbitrary PLA.ScriptContext where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary (PLA.ScriptContext txi purpose) =
+    coarbitrary txi . coarbitrary purpose
+
+-- | @since 1.0.0
+instance Function PLA.ScriptContext where
+  {-# INLINEABLE function #-}
+  function = functionMap (\(PLA.ScriptContext txi purpose) -> (txi, purpose)) (uncurry PLA.ScriptContext)
+
+-- | @since 1.0.0
+deriving via PLA.DatumHash instance Arbitrary PLA.RedeemerHash
+
+-- | @since 1.0.0
+deriving via PLA.DatumHash instance CoArbitrary PLA.RedeemerHash
+
+-- | @since 1.0.0
+instance Function PLA.RedeemerHash where
+  {-# INLINEABLE function #-}
+  function = functionMap coerce PLA.RedeemerHash
+
+{- | This is a very general instance, able to produce 'PLC.Data' of basically
+any shape. You probably want something more focused than this.
+
+@since 1.0.0
+-}
+instance Arbitrary PLC.Data where
```

Comment:
Any reason we have instances for both `Data` and `BuiltinData`

`BuiltinData` is just wrapper around `Data`

Reply:
Not exactly. `Data` doesn't exist onchain, while `BuiltinData` does, so having both does make some sense. It's minor, but since the two instances are basically identical, it doesn't really hurt too much to have both.